clients/spark/core/src/test/scala/io/treeverse/clients/ApiClientSpec.scala

```diff
@@ -32,17 +38,136 @@ class ApiClientSpec extends AnyFunSpec with Matchers {
+    @throws[Exception]
+    def someMethod(): String = {
+      SomeValue
+    }
+
+    def createCheckedSupplierForSomeMethod(): dev.failsafe.function.CheckedSupplier[String] = {
+      new dev.failsafe.function.CheckedSupplier[String]() {
+        def get(): String = someMethod()
+      }
+    }
+  }
+
+  describe("wrapWithRetry") {
+    // In case all retries fail, the last result or exception are returned or thrown
+    // https://failsafe.dev/faqs/#how-to-i-throw-an-exception-when-retries-are-exceeded
+    it(
+      "should throw the a FailSafeException exception in case all retries failed and last exception is checked"
+    ) {
+      // prepare
+      doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[ApiException])
+        .when(dummyMethodInvoker)
+        .someMethod()
+
+      assertThrows[FailsafeException] {
+        retryWrapper.wrapWithRetry(dummyMethodInvoker.createCheckedSupplierForSomeMethod())
+      }
+
+      // assert
+      verify(dummyMethodInvoker, times(DefaultMaxNumAttempts)).someMethod()
+    }
+
+    it("should throw the last unchecked exception in case all retries failed") {
+      // prepare
+      doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doThrow(classOf[NullPointerException])
+        .when(dummyMethodInvoker)
+        .someMethod()
+
+      assertThrows[NullPointerException] {
+        retryWrapper.wrapWithRetry(dummyMethodInvoker.createCheckedSupplierForSomeMethod())
+      }
+
+      // assert
+      verify(dummyMethodInvoker, times(DefaultMaxNumAttempts)).someMethod()
+    }
+
+    it("should retry on any type of exception") {
+      // prepare
+      doThrow(classOf[RuntimeException])
+        .doThrow(classOf[SocketException])
+        .doThrow(classOf[ApiException])
+        .doThrow(classOf[NullPointerException])
+        .doCallRealMethod()
+        .when(dummyMethodInvoker)
+        .someMethod()
+
+      // test
+      retryWrapper.wrapWithRetry(dummyMethodInvoker.createCheckedSupplierForSomeMethod())
+
+      // assert
+      verify(dummyMethodInvoker, times(5)).someMethod()
+    }
+
+    it("should retry until success within the boundaries of the allowed number of retries") {
+      // prepare
+      doThrow(classOf[Exception])
+        .doThrow(classOf[Exception])
+        .doCallRealMethod()
+        .when(dummyMethodInvoker)
+        .someMethod()
+
+      // test
+      retryWrapper.wrapWithRetry(dummyMethodInvoker.createCheckedSupplierForSomeMethod())
+
+      // assert
+      verify(dummyMethodInvoker, times(3)).someMethod()
+    }
+
+    it("should use custom max number of retries") {
+      retryWrapper = new RequestRetryWrapper(3)
+      assert(retryWrapper.numRetries == 3)
```

Comment:
Nit: This does not test the number of retries, it tests the constructor copies its parameter into some field.